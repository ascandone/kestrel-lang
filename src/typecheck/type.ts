export type ConcreteType =
  | {
      type: "fn";
      args: Type[];
      return: Type;
    }
  | {
      type: "named";
      moduleName: string;
      name: string;
      args: Type[];
    };

export type Type =
  | ConcreteType
  | {
      type: "var";
      var: TVar;
    };

export type TVarResolution =
  | { type: "unbound"; id: number }
  | { type: "bound"; value: ConcreteType };

export type UnifyError = {
  type: UnifyErrorType;
  left: Type;
  right: Type;
};

export class TVar {
  private constructor(
    private value: TVarResolution | { type: "linked"; to: TVar },
  ) {}

  static fresh(): TVar {
    return new TVar({ type: "unbound", id: TVar.unboundId++ });
  }

  resolve(): TVarResolution {
    if (this.value.type === "linked") {
      return this.value.to.resolve();
    }

    return this.value;
  }

  private static unboundId = 0;
  static resetId() {
    TVar.unboundId = 0;
  }

  asType(): Type & { type: "var" } {
    return { type: "var", var: this };
  }

  static unify(t1: Type, t2: Type): UnifyError | undefined {
    // (Var, Var)
    if (t1.type === "var" && t2.type === "var") {
      return TVar.unifyVars(t1.var, t2.var);
    }

    // (Var, _)
    if (t1.type === "var" && t2.type !== "var") {
      const occurs = occursCheck(t1.var, t2);
      if (occurs) {
        return {
          type: "occurs-check",
          left: t1.var.asType(),
          right: t2,
        };
      }

      switch (t1.var.value.type) {
        case "bound":
          return TVar.unify(t1.var.value.value, t2);
        case "unbound":
          t1.var.value = { type: "bound", value: t2 };
          return;
        case "linked":
          return TVar.unify(t1.var.value.to.asType(), t2);
        default:
          throw new Error("[unreachable]");
      }
    }

    // (_, Var)
    if (t2.type === "var") {
      return TVar.unify(t2, t1);
    }

    // (Named, Named)
    if (t1.type === "named" && t2.type === "named") {
      if (
        t1.name !== t2.name ||
        t1.moduleName !== t2.moduleName ||
        t1.args.length !== t2.args.length
      ) {
        return { type: "type-mismatch", left: t1, right: t2 };
      }

      for (let i = 0; i < t1.args.length; i++) {
        const res = TVar.unify(t1.args[i]!, t2.args[i]!);
        if (res !== undefined) {
          return res;
        }
      }

      return;
    }

    // (Fn, Fn)
    if (t1.type === "fn" && t2.type === "fn") {
      if (t1.args.length !== t2.args.length) {
        return { type: "type-mismatch", left: t1, right: t2 };
      }

      for (let i = 0; i < t1.args.length; i++) {
        const res = TVar.unify(t1.args[i]!, t2.args[i]!);
        if (res !== undefined) {
          return res;
        }
      }

      return TVar.unify(t1.return, t2.return);
    }

    // (_, _)
    return { type: "type-mismatch", left: t1, right: t2 };
  }

  private static unifyVars($1: TVar, $2: TVar): UnifyError | undefined {
    const r1 = $1.resolve();
    if (r1.type === "bound") {
      return TVar.unify($2.asType(), r1.value);
    }

    const r2 = $2.resolve();
    if (r2.type === "bound") {
      return TVar.unify($1.asType(), r2.value);
    }

    if ($2.value.type === "linked") {
      return TVar.unify($2.value.to.asType(), $1.asType());
    }

    if (r1.type === "unbound" && r2.type === "unbound" && r1.id === r2.id) {
      // TVars are already linked
      return;
    }

    $2.value = { type: "linked", to: $1 };
    return undefined;
  }
}

export type UnifyErrorType = "type-mismatch" | "occurs-check";

export const unify = TVar.unify;

// Occurs check on monotypes
function occursCheck(v: TVar, x: Type): boolean {
  if (x.type === "named") {
    return x.args.some((a) => occursCheck(v, a));
  }

  if (x.type === "fn") {
    return x.args.some((a) => occursCheck(v, a)) || occursCheck(v, x.return);
  }

  const resolvedV = v.resolve();

  if (resolvedV.type === "bound") {
    return false;
  }

  const resolvedX = x.var.resolve();

  if (resolvedX.type === "bound") {
    return false;
  }

  if (resolvedV.id === resolvedX.id) {
    return true;
  }

  return false;
}

export type Context = Record<string, Type>;

const FIRST_CHAR = "a".charCodeAt(0);
const LAST_CHAR = "z".charCodeAt(0);
const TOTAL_CHARS = LAST_CHAR - FIRST_CHAR + 1;

function generalizedName(count: number): string {
  const letter = String.fromCharCode(FIRST_CHAR + (count % TOTAL_CHARS));
  const rem = Math.floor(count / TOTAL_CHARS);

  if (rem === 0) {
    return letter;
  }

  return `${letter}${rem}`;
}

export type TypeScheme = Record<number, string>;

export type PolyType = [TypeScheme, Type];

export function generalizeAsScheme(
  mono: Type,
  initialScheme: TypeScheme = {},
): TypeScheme {
  const usedNames = new Set(Object.values(initialScheme));
  let nextId = 0;
  const scheme: TypeScheme = { ...initialScheme };

  function recur(mono: Type) {
    switch (mono.type) {
      case "var": {
        const res = mono.var.resolve();
        switch (res.type) {
          case "unbound": {
            if (res.id in scheme) {
              return;
            }

            // eslint-disable-next-line no-constant-condition
            while (true) {
              const name = generalizedName(nextId++);
              if (!usedNames.has(name)) {
                scheme[res.id] = name;
                break;
              }
            }

            return;
          }
          case "bound":
            recur(res.value);
            return;
        }
      }

      case "named":
        for (const arg of mono.args) {
          recur(arg);
        }
        return;

      case "fn":
        for (const arg of mono.args) {
          recur(arg);
        }
        recur(mono.return);
        return;
    }
  }

  recur(mono);
  return scheme;
}

export function instantiateFromScheme(mono: Type, scheme: TypeScheme): Type {
  const instantiated = new Map<string, TVar>();

  function recur(mono: Type): Type {
    switch (mono.type) {
      case "named":
        return {
          ...mono,
          args: mono.args.map(recur),
        };
      case "fn":
        if (mono.type !== "fn") {
          throw new Error("Invalid type");
        }

        return {
          type: "fn",
          args: mono.args.map(recur),
          return: recur(mono.return),
        };

      case "var": {
        const resolved = mono.var.resolve();
        switch (resolved.type) {
          case "unbound": {
            const boundId = scheme[resolved.id];
            if (boundId === undefined) {
              return mono;
            }

            const i = instantiated.get(boundId);
            if (i !== undefined) {
              return i.asType();
            }

            const t = TVar.fresh();
            instantiated.set(boundId, t);
            return t.asType();
          }
          case "bound":
            return recur(resolved.value);
        }
      }
    }
  }

  return recur(mono);
}

function typeToStringHelper(t: Type, scheme: TypeScheme): string {
  switch (t.type) {
    case "var": {
      const resolved = t.var.resolve();
      switch (resolved.type) {
        case "bound":
          return typeToStringHelper(resolved.value, scheme);
        case "unbound": {
          const id = scheme[resolved.id];
          if (id === undefined) {
            throw new Error("[unreachable] var not found: " + resolved.id);
          }
          return id;
        }
      }
    }

    case "fn": {
      const args = t.args
        .map((arg) => typeToStringHelper(arg, scheme))
        .join(", ");
      return `Fn(${args}) -> ${typeToStringHelper(t.return, scheme)}`;
    }

    case "named": {
      if (t.args.length === 0) {
        return t.name;
      }

      if (t.name === "Tuple2") {
        return `(${typeToStringHelper(t.args[0]!, scheme)}, ${typeToStringHelper(t.args[1]!, scheme)})`;
      }

      const args = t.args
        .map((arg) => typeToStringHelper(arg, scheme))
        .join(", ");
      return `${t.name}<${args}>`;
    }
  }
}

export function typeToString(t: Type, scheme?: TypeScheme): string {
  scheme = generalizeAsScheme(t, scheme);
  return typeToStringHelper(t, scheme);
}
