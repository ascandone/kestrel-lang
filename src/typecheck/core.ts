import { Type } from "./type";

export const bool: Type = {
  type: "named",
  moduleName: "Bool",
  name: "Bool",
  args: [],
};

export const int: Type = {
  moduleName: "Int",
  type: "named",
  name: "Int",
  args: [],
};

export const float: Type = {
  moduleName: "Float",
  type: "named",
  name: "Float",
  args: [],
};

export const string: Type = {
  moduleName: "String",
  type: "named",
  name: "String",
  args: [],
};

export const char: Type = {
  moduleName: "Char",
  type: "named",
  name: "Char",
  args: [],
};

export const unit: Type = {
  type: "named",
  moduleName: "Tuple",
  name: "Unit",
  args: [],
};

export const list = (a: Type): Type => ({
  type: "named",
  moduleName: "List",
  name: "List",
  args: [a],
});

export function task(arg: Type): Type {
  return {
    type: "named",
    moduleName: "Task",
    name: "Task",
    args: [arg],
  };
}
