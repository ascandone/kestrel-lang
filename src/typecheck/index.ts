export {
  typecheck,
  typecheckProject,
  UnifyErrorType,
  TypecheckError,
  ProjectTypeCheckResult,
} from "./typecheck";
export { typeErrorPPrint, typePPrint } from "./pretty-printer";
