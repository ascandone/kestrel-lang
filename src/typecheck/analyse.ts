/* eslint-disable require-yield */
import {
  ErrorInfo,
  InvalidPipe,
  OccursCheck,
  TraitNotSatified,
  TypeMismatch,
  UnboundVariable,
  UnusedVariable,
} from "../errors";
import {
  Binding,
  ConstLiteral,
  RangeMeta,
  TypeAst,
  UntypedDeclaration,
  UntypedExpr,
  UntypedImport,
  UntypedMatchPattern,
  UntypedModule,
  UntypedTypeDeclaration,
  UntypedTypeVariant,
  parse,
} from "../parser";
import { bool, char, float, int, list, string } from "./core";
import { Deps } from "./resolutionStep";
import { PolyType, TVar, Type, UnifyError, instantiate, unify } from "./type";

export type AnalyseOptions = {
  dependencies?: Deps;
  implicitImports?: UntypedImport[];
  mainType?: Type;
};

export type TypedNode = Binding | UntypedExpr;
export type IdentifierResolution =
  | {
      type: "local-variable";
      binding: Binding;
    }
  | {
      type: "global-variable";
      declaration: UntypedDeclaration;
      namespace: string;
    }
  | {
      type: "constructor";
      variant: UntypedTypeVariant;
      declaration: UntypedTypeDeclaration & { type: "adt" };
      namespace: string;
    };

type LocalScope = Record<string, Binding>;

export class Analysis {
  errors: ErrorInfo[] = [];

  /** record of types declared in this module */
  private locallyDefinedTypes = new Map<string, UntypedTypeDeclaration>();

  /** record of variants declared in this module */
  private locallyDefinedVariants = new Map<
    string,
    [UntypedTypeDeclaration & { type: "adt" }, UntypedTypeVariant]
  >();

  private identifiersResolutions = new WeakMap<
    UntypedExpr & { type: "identifier" },
    IdentifierResolution
  >();
  private unusedBindings = new WeakSet<Binding>();

  private typeAnnotations = new WeakMap<TypedNode, TVar>();
  private module: UntypedModule;

  constructor(
    public readonly ns: string,
    public readonly source: string,
    // private options: AnalyseOptions = {},
  ) {
    const parseResult = parse(source);
    // TODO push parsing/lexer errs in errs

    this.module = parseResult.parsed;

    this.initTypesResolution();
    this.initDeclarationsResolution();

    this.initDeclarationsTypecheck();
  }

  private initTypesResolution() {
    // 1. register all the types
    for (const typeDecl of this.module.typeDeclarations) {
      this.locallyDefinedTypes.set(typeDecl.name, typeDecl);
    }

    // 2. register constructors and fields
    for (const typeDecl of this.module.typeDeclarations) {
      switch (typeDecl.type) {
        case "adt":
          for (const variant of typeDecl.variants) {
            this.locallyDefinedVariants.set(variant.name, [typeDecl, variant]);
          }
          break;
        case "struct":
          throw new Error("TODO handle struct");
        case "extern":
          break;
      }
    }
  }

  private initDeclarationsResolution() {
    for (const letDecl of this.module.declarations) {
      if (letDecl.extern) {
        continue;
      }

      this.runValuesResolution(letDecl.value, {});
    }
  }

  private initDeclarationsTypecheck() {
    for (const letDecl of this.module.declarations) {
      this.typecheckLetDeclaration(letDecl);
    }
  }

  private typecheckLetDeclaration(decl: UntypedDeclaration) {
    if (decl.typeHint !== undefined) {
      const typeHintType = this.typeAstToType(decl.typeHint.mono);
      this.unifyNode(decl.binding, typeHintType);
    }
    if (decl.extern) {
      return;
    }

    this.typecheckExpr(decl.value);
    this.unifyNodes(decl.binding, decl.value);
  }

  private unifyNode(node: TypedNode, type: Type) {
    const err = unify(this.getType(node), type);
    if (err !== undefined) {
      this.errors.push(unifyErrToErrorInfo(node, err));
      return;
    }
  }

  private unifyNodes(left: TypedNode, right: TypedNode) {
    this.unifyNode(left, this.getType(right));
  }

  private evaluateResolution(
    identifier: UntypedExpr & { type: "identifier" },
    localScope: LocalScope = {},
  ): IdentifierResolution | undefined {
    // Search locals first
    const localLookup = localScope[identifier.name];
    if (localLookup !== undefined) {
      return { type: "local-variable", binding: localLookup };
    }

    // search variants
    const variantLookup = this.locallyDefinedVariants.get(identifier.name);
    if (variantLookup !== undefined) {
      const [declaration, variant] = variantLookup;

      return {
        type: "constructor",
        namespace: this.ns,
        declaration,
        variant,
      };
    }

    // TODO search locallyDefinedDeclarations instead
    for (const declaration of this.getDeclarations()) {
      if (declaration.binding.name === identifier.name) {
        return {
          type: "global-variable",
          declaration,
          namespace: this.ns,
        };
      }
    }
    return undefined;
  }

  private extractPatternIdentifiers(pattern: UntypedMatchPattern): Binding[] {
    switch (pattern.type) {
      case "identifier":
        return [pattern];

      case "lit":
      case "constructor":
        throw new Error("TODO handle pattern of type: " + pattern.type);
    }
  }

  private checkUnusedVars(expr: Binding) {
    if (this.unusedBindings.has(expr)) {
      this.errors.push({
        range: expr.range,
        description: new UnusedVariable(expr.name, "local"),
      });
    }
  }

  private runValuesResolution(
    expr: UntypedExpr,
    localScope: LocalScope,
  ): undefined {
    switch (expr.type) {
      case "syntax-err":
      case "constant":
        return;

      case "identifier": {
        const res = this.evaluateResolution(expr, localScope);
        if (res === undefined) {
          this.errors.push({
            range: expr.range,
            description: new UnboundVariable(expr.name),
          });
          return;
        }

        this.identifiersResolutions.set(expr, res);
        switch (res.type) {
          case "local-variable":
            this.unusedBindings.delete(res.binding);
            break;

          case "global-variable":
          case "constructor":
            break;
        }

        return;
      }

      case "fn": {
        const paramsBindings = expr.params
          .flatMap((p) => this.extractPatternIdentifiers(p))
          .map((p) => [p.name, p]);

        this.runValuesResolution(expr.body, {
          ...localScope,
          ...Object.fromEntries(paramsBindings),
        });
        return;
      }

      case "application": {
        this.runValuesResolution(expr.caller, localScope);
        for (const arg of expr.args) {
          this.runValuesResolution(arg, localScope);
        }

        return;
      }

      case "pipe": {
        this.runValuesResolution(expr.right, localScope);
        this.runValuesResolution(expr.left, localScope);
        return;
      }

      case "if": {
        this.runValuesResolution(expr.condition, localScope);
        this.runValuesResolution(expr.then, localScope);
        this.runValuesResolution(expr.else, localScope);
        return;
      }

      case "let": {
        // TODO handle recursive bindings
        this.runValuesResolution(expr.value, localScope);

        const bindingsEntries = this.extractPatternIdentifiers(
          expr.pattern,
        ).map((p) => [p.name, p] as const);
        for (const [_, binding] of bindingsEntries) {
          this.unusedBindings.add(binding);
        }

        this.runValuesResolution(expr.body, {
          ...localScope,
          ...Object.fromEntries(bindingsEntries),
        });

        for (const [_, binding] of bindingsEntries) {
          this.checkUnusedVars(binding);
        }
        return;
      }

      case "list-literal": {
        for (const value of expr.values) {
          this.runValuesResolution(value, localScope);
        }
        return;
      }

      case "let#":
      case "infix":
      case "struct-literal":
      case "field-access":
      case "match":
        throw new Error("TODO resolution on: " + expr.type);
    }
  }

  private typecheckExpr(expr: UntypedExpr): undefined {
    switch (expr.type) {
      case "syntax-err":
        return;

      case "constant":
        this.unifyNode(expr, getConstantType(expr.value));
        return;

      case "identifier": {
        const resolution = this.identifiersResolutions.get(expr);
        if (resolution === undefined) {
          // error was already emitted during resolution
          return;
        }

        switch (resolution.type) {
          case "global-variable":
            // TODO instantiate
            // TODO check order
            this.unifyNodes(expr, resolution.declaration.binding);
            return;

          case "local-variable":
            this.unifyNodes(expr, resolution.binding);
            return;

          case "constructor": {
            const mono = instantiate(
              this.getVariantType(resolution.variant, resolution.declaration),
            );
            this.unifyNode(expr, mono);
            return;
          }
        }
      }

      case "fn":
        this.unifyNode(expr, {
          type: "fn",
          args: expr.params.map((pattern) => {
            if (pattern.type !== "identifier") {
              throw new Error("handle pattern != ident");
            }
            return this.getType(pattern);
          }),
          return: this.getType(expr.body),
        });
        this.typecheckExpr(expr.body);
        return;

      case "application":
        this.typecheckExpr(expr.caller);
        this.unifyNode(expr.caller, {
          type: "fn",
          args: expr.args.map((arg) => this.getType(arg)),
          return: this.getType(expr),
        });
        for (const arg of expr.args) {
          this.typecheckExpr(arg);
        }
        return;

      case "if":
        this.unifyNode(expr.condition, bool);
        this.unifyNodes(expr, expr.then);
        this.unifyNodes(expr, expr.else);

        this.typecheckExpr(expr.condition);
        this.typecheckExpr(expr.then);
        this.typecheckExpr(expr.else);
        return;

      case "let":
        this.unifyNodes(expr, expr.value);

        this.typecheckExpr(expr.body);
        this.typecheckExpr(expr.value);
        return;

      case "list-literal": {
        const listType = TVar.fresh().asType();
        this.unifyNode(expr, list(listType));
        for (const value of expr.values) {
          this.unifyNode(value, listType);
          this.typecheckExpr(value);
        }
        return;
      }

      case "pipe": {
        if (expr.right.type !== "application") {
          this.errors.push({
            range: expr.right.range,
            description: new InvalidPipe(),
          });
          return;
        }

        // Do not typecheck 'expr.right' recursively: function has wrong number of args
        this.typecheckExpr(expr.right.caller);
        this.typecheckExpr(expr.left);
        this.unifyNode(expr.right.caller, {
          type: "fn",
          args: [
            this.getType(expr.left),
            ...expr.right.args.map((arg) => this.getType(arg)),
          ],
          return: this.getType(expr),
        });
        return;
      }

      case "let#":
      case "infix":
      case "struct-literal":
      case "field-access":
      case "match":
        throw new Error("TODO handle typecheck of: " + expr.type);
    }
  }

  private typeAstToType(t: TypeAst): Type {
    const helper = (t: TypeAst): Type => {
      switch (t.type) {
        case "named":
          // TODO actually resolve type
          return {
            type: "named",
            args: t.args.map((arg) => helper(arg)),
            moduleName: this.ns,
            name: t.name,
          };

        case "fn":
          return {
            type: "fn",
            args: t.args.map((arg) => helper(arg)),
            return: helper(t.return),
          };

        case "var":
          // TODO
          return TVar.fresh().asType();

        case "any":
          return TVar.fresh().asType();
      }
    };

    return helper(t);
  }

  private getVariantType(
    variant: UntypedTypeVariant,
    declaration: UntypedTypeDeclaration & { type: "adt" },
  ): PolyType {
    // TODO cache and generalize

    const ret: Type = {
      type: "named",
      args: [],
      moduleName: this.ns,
      name: declaration.name,
    };

    if (variant.args.length === 0) {
      return [{}, ret];
    }

    return [
      {},
      {
        type: "fn",
        args: variant.args.map((arg) => this.typeAstToType(arg)),
        return: ret,
      },
    ];
  }

  // --- Public interface
  getType(node: TypedNode): Type {
    const lookup = this.typeAnnotations.get(node);
    if (lookup === undefined) {
      // initialize node
      const tvar = TVar.fresh();
      this.typeAnnotations.set(node, tvar);
      return tvar.asType();
    }
    return lookup.asType();
  }

  *getDeclarations(): Generator<UntypedDeclaration> {
    for (const decl of this.module.declarations) {
      yield decl;
    }
  }

  *getPublicDeclarations(): Generator<UntypedDeclaration> {
    for (const decl of this.getDeclarations()) {
      if (decl.pub) {
        yield decl;
      }
    }
  }
}

function unifyErrToErrorInfo(node: RangeMeta, e: UnifyError): ErrorInfo {
  switch (e.type) {
    case "missing-trait":
      return {
        range: node.range,
        description: new TraitNotSatified(e.type_, e.trait),
      };

    case "type-mismatch":
      return {
        range: node.range,
        description: new TypeMismatch(e.left, e.right),
      };

    case "occurs-check":
      return { range: node.range, description: new OccursCheck() };
  }
}

// Keep this in sync with core
function getConstantType(x: ConstLiteral): Type {
  // Keep this in sync with core
  switch (x.type) {
    case "int":
      return int;

    case "float":
      return float;

    case "string":
      return string;

    case "char":
      return char;
  }
}
