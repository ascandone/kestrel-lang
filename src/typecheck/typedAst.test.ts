import { test, expect, describe } from "vitest";
import {
  Hovered,
  goToDefinitionOf,
  hoverOn,
  hoverToMarkdown,
  Location,
} from "./typedAst";
import { Span, unsafeParse } from "../parser";
import { typecheck } from "./typecheck";
import { TypeScheme } from "./type";

describe("hoverOn", () => {
  test("hover a declaration's binding", () => {
    const src = `let x = 42`;
    const [, hoverable] = parseHover(src, "x")!;
    expect(hoverable).toEqual<Hovered>({
      span: spanOf(src, "x"),
      hovered: expect.objectContaining({
        type: "global-variable",
        declaration: expect.objectContaining({
          binding: expect.objectContaining({
            name: "x",
          }),
        }),
      }),
    });
  });

  test("hover a declaration's binding (2d occurrence)", () => {
    const src = `let x = 42\nlet y = 100`;
    const [, hoverable] = parseHover(src, "y")!;
    expect(hoverable).toEqual<Hovered>({
      span: spanOf(src, "y"),
      hovered: expect.objectContaining({
        type: "global-variable",
        declaration: expect.objectContaining({
          binding: expect.objectContaining({
            name: "y",
          }),
        }),
      }),
    });
  });

  test("hover fn param", () => {
    const src = `let f = fn x { 0 }`;
    const [, hoverable] = parseHover(src, "x")!;
    expect(hoverable).toEqual<Hovered>({
      span: spanOf(src, "x"),
      hovered: expect.objectContaining({
        type: "local-variable",
        binding: expect.objectContaining({
          name: "x",
        }),
      }),
    });
  });

  test("hover let expr binding", () => {
    const src = `let f = {
      let x = 42;
      0
    }`;
    const [, hoverable] = parseHover(src, "x")!;
    expect(hoverable).toEqual<Hovered>({
      span: spanOf(src, "x"),
      hovered: expect.objectContaining({
        type: "local-variable",
        binding: expect.objectContaining({
          name: "x",
        }),
      }),
    });
  });

  test("hovering with scheme", () => {
    const src = `let f = fn x, y { y }`;
    const [scheme, hoverable] = parseHover(src, "y", 2)!;
    if (hoverable.hovered.type !== "local-variable") {
      throw new Error("fail");
    }
    const resolved = hoverable.hovered.binding.$.resolve();
    if (resolved.type !== "unbound") {
      throw new Error("fail");
    }
    expect(scheme).toEqual(
      expect.objectContaining({
        [resolved.id]: "b",
      }),
    );
  });

  test("hover a reference to a global binding", () => {
    const src = `
        let x = 42
        let y = x
    `;
    const [, hoverable] = parseHover(src, "x", 2)!;
    expect(hoverable).toEqual<Hovered>({
      span: spanOf(src, "x", 2),
      hovered: expect.objectContaining({
        type: "global-variable",
        declaration: expect.objectContaining({
          binding: expect.objectContaining({
            name: "x",
          }),
        }),
      }),
    });
  });

  test("hover a reference to a global binding inside a fn, application, if, let", () => {
    const src = `
        let x = 42
        let y = fn {
            if true {
                let y = 0;
                f(0, x)
            } else {
                0
            }
        }
    `;
    const [, hoverable] = parseHover(src, "x", 2)!;
    expect(hoverable).toEqual<Hovered>({
      span: spanOf(src, "x", 2),
      hovered: expect.objectContaining({
        type: "global-variable",
        declaration: expect.objectContaining({
          binding: expect.objectContaining({
            name: "x",
          }),
        }),
      }),
    });
  });

  test("hover a local binding in a match expr", () => {
    const src = `
        let m = match something {
          Constr(Nested(_, loc_var)) => 0,
        }
    `;
    const [, hoverable] = parseHover(src, "loc_var")!;
    expect(hoverable).toEqual<Hovered>(
      expect.objectContaining({
        span: spanOf(src, "loc_var"),
      }),
    );
  });

  test("hover a reference to a local binding in a match expr", () => {
    const src = `
        let m = match something {
            0 => 0,
            Constr(x) => x,
        }
    `;
    const [, hoverable] = parseHover(src, "x", 2)!;
    expect(hoverable).toEqual<Hovered>({
      span: spanOf(src, "x", 2),
      hovered: expect.objectContaining({
        type: "local-variable",
        binding: expect.objectContaining({
          name: "x",
        }),
      }),
    });
  });

  test("hover a reference to a local binding in a match expr subject", () => {
    const src = `
      let something = 42
      let m = match something { }
    `;

    const [, hoverable] = parseHover(src, "something", 2)!;
    expect(hoverable).toEqual<Hovered>(
      expect.objectContaining({
        span: spanOf(src, "something", 2),
      }),
    );
  });

  test("hover a constructor", () => {
    const src = `
        type T { Constr(Int) }
        let t = Constr(42)
    `;
    const [, hoverable] = parseHover(src, "Constr", 2)!;
    expect(hoverable).toEqual<Hovered>({
      span: spanOf(src, "Constr", 2),
      hovered: expect.objectContaining({
        type: "constructor",
      }),
    });
  });

  test("snapshot when hovering on global fn", () => {
    const src = `let glb = fn x, y { y }`;
    const [scheme, hoverable] = parseHover(src, "glb")!;
    expect(hoverToMarkdown(scheme, hoverable)).toMatchSnapshot();
  });

  test("snapshot when hovering on an extern type", () => {
    const src = `extern let glb: Fn(x) -> x`;
    const [scheme, hoverable] = parseHover(src, "glb")!;
    expect(hoverToMarkdown(scheme, hoverable)).toMatchSnapshot();
  });

  test("snapshot when hovering on an extern type reference", () => {
    const src = `
      extern let glb: Fn(x) -> x
      let v = glb
    `;
    const [scheme, hoverable] = parseHover(src, "glb", 2)!;
    expect(hoverToMarkdown(scheme, hoverable)).toMatchSnapshot();
  });

  test("snapshot when hovering on a local fn", () => {
    const src = `
      let glb = fn a {
        let closure = fn x { a };
        0
      }
    `;
    const [scheme, hoverable] = parseHover(src, "closure")!;
    expect(hoverToMarkdown(scheme, hoverable)).toMatchSnapshot();
  });

  test("snapshot when hovering on a constructor", () => {
    const src = `
      type T<content> { X, Constr(content) }
    `;
    const [scheme, hoverable] = parseHover(src, "Constr")!;
    expect(hoverToMarkdown(scheme, hoverable)).toMatchSnapshot();
  });

  test("snapshot when hovering on a constructor reference", () => {
    const src = `
      type T<content> { Constr(content) }

      pub let c = Constr(42)
    `;
    const [scheme, hoverable] = parseHover(src, "Constr", 2)!;
    expect(hoverToMarkdown(scheme, hoverable)).toMatchSnapshot();
  });
});

describe("goToDefinition", () => {
  test("glb decl", () => {
    const src = `
      let glb = 42
      let _ = glb
    `;
    const location = parseGotoDef(src, "glb", 2);
    expect(location?.span).toEqual(spanOf(src, "glb"));
  });

  test("let decl", () => {
    const src = `let _ = { let x = 42; 1 + x }`;
    const location = parseGotoDef(src, "x", 2);
    expect(location?.span).toEqual(spanOf(src, "x", 1));
  });

  test("fn par", () => {
    const src = `let _ = fn x { x }`;
    const location = parseGotoDef(src, "x", 2);
    expect(location?.span).toEqual(spanOf(src, "x", 1));
  });

  test("wrapped in an if and appl", () => {
    const src = `let _ = fn a, loc_var {
      if True {
        0
      } else {
        f(loc_var)
      }
    }`;
    const location = parseGotoDef(src, "loc_var", 2);
    expect(location?.span).toEqual(spanOf(src, "loc_var", 1));
  });

  test("pattern ident", () => {
    const src = `
      type Box<a> { Box(a) }
      let _ = match Box(Box(42)) {
        _ => 0,
        Box(Box(local_var)) => local_var,
      }
    `;
    const location = parseGotoDef(src, "local_var", 2);
    expect(location?.span).toEqual(spanOf(src, "local_var", 1));
  });

  test("valid recursive let bindings", () => {
    const src = `
      let x = f(fn { x })
    `;
    const location = parseGotoDef(src, "x", 2);
    expect(location?.span).toEqual(spanOf(src, "x", 1));
  });

  test("invalid recursive let bindings", () => {
    const src = `
      let x = x
    `;
    const location = parseGotoDef(src, "x", 2);
    expect(location?.span).toEqual(undefined);
  });

  test("shadowed let", () => {
    const src = `
      let x = {
        let a = 0;
        let a = a;
        42
      }
    `;
    const location = parseGotoDef(src, "a", 3);
    expect(location?.span).toEqual(spanOf(src, "a", 1));
  });
});

function parseHover(
  src: string,
  hovering: string,
  occurrenceNumber = 1,
): [TypeScheme, Hovered] | undefined {
  const offset = indexOf(src, hovering, occurrenceNumber);

  if (offset === undefined) {
    throw new Error("Invalid offset");
  }

  const parsed = unsafeParse(src);
  const [typed, _] = typecheck("Main", parsed);
  return hoverOn("Ns", typed, offset);
}

function parseGotoDef(
  src: string,
  hovering: string,
  occurrenceNumber = 1,
): Location | undefined {
  const offset = indexOf(src, hovering, occurrenceNumber);

  if (offset === undefined) {
    throw new Error("Invalid offset");
  }

  const parsed = unsafeParse(src);
  const [typed, _] = typecheck("Main", parsed);
  return goToDefinitionOf(typed, offset);
}

function spanOf(
  src: string,
  subStr: string = src,
  occurrenceNumber: number = 1,
): Span {
  const index = indexOf(src, subStr, occurrenceNumber);
  if (index === undefined) {
    throw new Error("Invalid index");
  }

  return [index, index + subStr.length];
}

function indexOf(
  src: string,
  subStr: string = src,
  occurrenceNumber: number = 1,
): number | undefined {
  let lastOccurrenceIndex = -1;
  for (let i = 0; i < occurrenceNumber; i++) {
    lastOccurrenceIndex = src.indexOf(subStr, lastOccurrenceIndex + 1);
    if (lastOccurrenceIndex === -1) {
      return undefined;
    }
  }
  return lastOccurrenceIndex;
}

test("indexOf test utility", () => {
  expect(indexOf("a-a-a", "a")).toBe(0);
  expect(indexOf("a-a-a", "a", 2)).toBe(2);
  expect(indexOf("a-a-a", "a", 3)).toBe(4);
  expect(indexOf("a-a-a", "a", 4)).toBe(undefined);

  expect(indexOf("a-a-a", "not-found")).toBe(undefined);
  expect(indexOf("a-a-a", "not-found", 1)).toBe(undefined);

  expect(indexOf("ab-ab-ab", "ab")).toBe(0);
  expect(indexOf("ab-ab-ab", "ab", 2)).toBe(3);
  expect(indexOf("ab-ab-ab", "ab", 3)).toBe(6);

  expect(indexOf("let x =\n x + 1", "x", 2)).toBe(9);
});
