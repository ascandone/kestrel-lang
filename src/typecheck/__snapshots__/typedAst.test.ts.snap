// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`hoverOn > snapshot when hovering on a constructor 1`] = `
"\`\`\`
Constr: Fn(content) -> T<content>
\`\`\`
type constructor
"
`;

exports[`hoverOn > snapshot when hovering on a constructor reference 1`] = `
"\`\`\`
Constr: Fn(content) -> T<content>
\`\`\`
type constructor
"
`;

exports[`hoverOn > snapshot when hovering on a local fn 1`] = `
"\`\`\`
closure: Fn(b) -> a
\`\`\`
local declaration
"
`;

exports[`hoverOn > snapshot when hovering on an extern type 1`] = `
"\`\`\`
glb: Fn(x) -> x
\`\`\`
global declaration
"
`;

exports[`hoverOn > snapshot when hovering on an extern type reference 1`] = `
"\`\`\`
glb: Fn(x) -> x
\`\`\`
global declaration
"
`;

exports[`hoverOn > snapshot when hovering on global fn 1`] = `
"\`\`\`
glb: Fn(a, b) -> b
\`\`\`
global declaration
"
`;
